### ISSUE TITLE
Adjoint models

### ISSUE THREAD

**Issue by RPrudden** [CliMT/climt#107]
Has anyone given any thought to calculating the adjoints of CliMT models? As far as I can tell there is no easy way to do this, but I'd be delighted to be wrong.

**Comment by mcgibbon** [comment]
I don't know much about adjoints, but if calculating an adjoint can be done by computing the outputs of [sympl](https://github.com/mcgibbon/sympl) components with slight perturbations to the inputs about some reference state, then I would say there should be an easy way to do it, but there isn't an existing way to do it. I don't know that anyone has done this before, maybe @JoyMonteiro has some insight. I think the result would make a neat Python package!

**Comment by JoyMonteiro** [comment]
@RPrudden you are correct that this is a non-trivial task. A lot of code in climt is in Fortran, and
there aren't really any good free tools (that I know of) that can do automatic differentiation (AD) of Fortran code. 

Even if we did get our hands on code like this, ADs have a hard time dealing with model physics because of strong nonlinearities present in the atmosphere in the presence of clouds and water vapour. See [this paper](https://rmets.onlinelibrary.wiley.com/doi/epdf/10.1002/qj.200212858306) for example.

While it would certainly be interesting to see adjoint models for climt components, it is not on my priority list currently.

If you are interested, I think the dynamical core would be a good target to develop an adjoint for.
The core dynamics is contained in a few files that make it quite amenable for such an effort.

**Comment by RPrudden** [comment]
Many thanks, both! I think computing the adjoint of the dynamical core would be enough to cover the kind of work I have in mind, so I may be able to avoid the difficulties involved in differentiating the physics components.

In terms of the dynamical core, it looks like [this directory](https://github.com/CliMT/climt/tree/develop/climt/_components/gfs) would be the place to start. I guess you were thinking to use [autograd](https://github.com/HIPS/autograd) to compute the gradients, is that right?

**Comment by JoyMonteiro** [comment]
That directory only contains the wrapper. The actual algorithmic part is written in Fortran and resides
[here](https://github.com/CliMT/climt/tree/develop/climt/_lib/GFS)

The easiest way I would think is to translate the relevant parts to python code and then run autograd on it. There are only two core functions:
* advance
https://github.com/CliMT/climt/blob/ddc6530a754d6778634b65ce3b93c9fabdf1f799/climt/_lib/GFS/run.f90#L72

which does the actual time stepping and
* getdyntend
https://github.com/CliMT/climt/blob/ddc6530a754d6778634b65ce3b93c9fabdf1f799/climt/_lib/GFS/dyn_run.f90#L140

which calculates the tendencies used by the timestepper. Each will need some helper functions,
but these are your entry points to the code.

There is a subtle issue to keep in mind -- There are parts of the code which do differentiation themselves, calling the shtns library. These will have to be translated by hand to calls to differentiation by shtns (there is a python wrapper to shtns, so not a big issue there).

I think it is doable, but requires some grunt work!

**Comment by RPrudden** [comment]
Ok, I understand better now. As you say, it looks fairly intensive to translate.. but I am quite interested, so I might try to carve out some time.

**Comment by JoyMonteiro** [comment]
Great!! We can keep this issue open in case you need any other input. While I can't contribute to the coding effort itself, I can certainly take a look at any code or help you to navigate the fortran code.
It would be really cool if this works out :)

**Comment by JoyMonteiro** [comment]
A bit of googling around yielded a AD package for Fortran:
http://www-sop.inria.fr/tropics/tapenade.html
which seems to be actively developed.

They seem to have [used it](http://www-sop.inria.fr/tropics/papers/LesHouches2012.pdf) for an ocean model (NEMO).
